### ISSUE TITLE
Host groups, Token groups, user groups

### ISSUE THREAD

**Issue by cornelinux** [privacyidea/privacyidea#285]
we need to create a concept of host groups, token groups and user groups.

This concept could be used to ease the assignment of tokens to machines.
This way token groups could be assigned to machine groups. 
This might help a lot with ssh key management.
The goal is not to create an assignment for *each* machine but *one* assignment for the whole *group*. If there are 100 SSH server where a the same administrator (or group of administrators) should be able to login with their SSH keys, the administrator should not assign 100 single machines but only *one* group (which contains all the machines).

User groups might already be implemented by the resolvers and realms.
# Host groups

host groups could be read from the host store like the AD.

Hosts with a filter in a certain attribute could be mapped togeather to a group. So should be able to define several filters for a host resolver. A filter needs to contain:
- identifier of the host group
- the filter expression, which is dependent of the host resolver type.

Thus, the filtering should be defined in the host resolver module.

There needs to be an additional database table, which connects:
- host group identifier
- host resolver
- filter 
## MachineResolver as host group

We could also use machine resolver as such hostgroups. Copies of machine resolvers are simple to create. This way e.g. in an LDAP host groups can be simply created by creating new machine resolver by slightly modifying the DN or search string.

Tokens than can be either assigned to a host or a host-group/host-resolver.
# Token groups

Just like we can assign tokens to realms we could assign tokens to token groups.
- Should tokens be assigned token by token to a group? That is tedious.
- We could choose a token group during enrollment.
- We could also define filters for tokens, to match these tokens into the token group. Hm. But this would break with named tokens in a group.

#2712

**Comment by cornelinux** [comment]
# Assign one token to a group of machines

In a first step we would need to be able to assign *one* token to a group of machines.

What needs to be done:

* [ ] Allow to group machines. Several resolvers need to be grouped into a machine-group (similar to user realms). F
  For this we need two new database tables ``machinegroup`` and ``machinegroupresolver``.
  **machinegroup**:

  * id
  * name

  **machinegroupresolver**:

  * group_id
  * machineresolver_id

* [ ] Allow to assign the token to a machine_group, not only to a machine. Or we totally change this an only allow to assign a token to a machine_group instead of a machine. After all, if you would need to assign a token to only *one* machine, you could still create a group with this single machine. We could enhance the table ``machinetoken`` with a column ``machine_group_id`` so we could either assign to a single machine or to a group.


# Assign group of tokens

Grouping tokens and assigning them to a machine or a group of machine can be left for later.

**Comment by B3DTech** [comment]
This sounds fantastic.  You're right - the need for a single machine assignment of a token is probably pretty limited, but I don't know if creating a group would be the best option.  It would be one more thing to clean up if/when the user doesn't need access/leaves the company/etc.  

A quick audit report of User to Machine mappings would be nice too.

E.g.:

Give me all machines user X is allowed to log in to.
Give me all users allowed to log into Machine X.  

What happens if a machine is somehow in multiple groups?

**Comment by jmozd** [comment]
> Grouping tokens and assigning them to a machine or a group of machine can be left for later.

I don't see the lower priority. It'd be nice to have an implicit group "all tokens of a user", ideally "all user's tokens of type x" usable where we can currently assign only single tokens - we already have the concept of users adding their own tokens, including a policy limiting the number of token per type. That way, a user may self-service i.e. the user's ssh keys, without having the need for an admin to assign the new tokens to where the old tokens were originally assigned.

If this were unwanted by the admins, they still could resort to adding individual tokens instead.

> It would be one more thing to clean up if/when the user doesn't need access/leaves the company/etc

I stumbled over the "referential integrity" constraints when trying to delete a user token (as the user owning the token) - in terms of "self service", a "cascaded delete" (probably at some API level, rather than at the SQL level) seems desirable.

**Comment by cornelinux** [comment]
User groups and Token groups could be used generically.

User groups e.g. could be used in policies or 4-eyes token.

Token groups e.g. could also be used in policies?

A group membership of a user actually is an attribute of the user. User attributes are requested in #680

**Comment by cornelinux** [comment]
This could be dependent on #680. We should first gather experience with #680 when v3.6 is productive. Then continue defining and working on how to use this for user groups or host groups. Moving this to 3.7.

**Comment by cornelinux** [comment]
Maybe in the future this could be done via policies.

**Comment by cornelinux** [comment]
This is superseded by #3299 and #3300.
Also see #2712
